File: java/AlgorithmAndDataStructure/src/test/java/algorithms/SortingTest.java
package algorithms;

import junit.framework.TestCase;

import java.util.*;

public class SortingTest extends TestCase {
	private ArrayList<Integer> a;
	private ArrayList<Integer> sorted;

	public void setUp() throws Exception {
		super.setUp();
		a = new ArrayList<Integer>(Arrays.asList(3, 1, 2, 4));
		sorted = new ArrayList<Integer>(Arrays.asList(1, 2, 3));
	}

	public void testIsSorted() throws Exception {
		assert !Sorting.isSorted(a);
		assert Sorting.isSorted(sorted);
	}

	public void testIsSortedUntil() throws Exception {
		assertEquals(1, Sorting.isSortedUntil(a).nextIndex());
	}

	public void testSort() throws Exception {
		Sorting.sort(a);
		assert Sorting.isSorted(a);
	}

	public void testPartialSort() throws Exception {
		Sorting.partialSort(a, 2);
		ArrayList<Integer> partialSorted = new ArrayList<Integer>(Arrays
				.asList(1, 2));
		assertEquals(partialSorted, a.subList(0, 2));
	}

	public void testPartialSortCopy() throws Exception {
		ArrayList<Integer> b = new ArrayList<Integer>();
		Sorting.partialSortCopy(a, 2, b);
		ArrayList<Integer> partialSorted = new ArrayList<Integer>(Arrays
				.asList(1, 2));
		assertEquals(partialSorted, b);
	}

	public void testNthElement() throws Exception {
		assertEquals(Integer.valueOf(2), Sorting.nthElement(a, 2));
	}

	public void testStableSort() throws Exception {
		ArrayList list = new ArrayList();
		list.add(new AbstractMap.SimpleEntry<Integer, Character>(1, 'b'));
		list.add(new AbstractMap.SimpleEntry<Integer, Character>(3, 'a'));
		list.add(new AbstractMap.SimpleEntry<Integer, Character>(1, 'a'));
		list.add(new AbstractMap.SimpleEntry<Integer, Character>(4, 'c'));
		Sorting.stableSort(list, new Comparator<AbstractMap
				.SimpleEntry<Integer, Character>>() {
			@Override
			public int compare(AbstractMap.SimpleEntry<Integer, Character> o1,
					AbstractMap.SimpleEntry<Integer, Character> o2) {
				return o1.getKey().compareTo(o2.getKey());
			}
		});
		assertEquals(new AbstractMap.SimpleEntry<Integer, Character>(1, 'b'),
				list.get(0));
		assertEquals(new AbstractMap.SimpleEntry<Integer, Character>(1, 'a'),
				list.get(1));
	}
}

File: php/SortingTest.php
<?php
namespace Algorithm;

class SortingTest extends \PHPUnit_Framework_TestCase
{
    private $ary;

    public function setUp() {
        $this->ary = [4, 7, 8, 1, 2];
    }

    public function testIsSortedUntil()
    {
        $this->assertEquals(3, isSortedUntil($this->ary));
        $this->assertEquals(1, isSortedUntil($this->ary, function($a, $b) {
            return $a >= $b;
        }));
        $this->assertNull(isSortedUntil([1, 5, 9]));
    }

    public function testIsSorted()
    {
        $this->assertFalse(isSorted($this->ary));
        $this->assertTrue(isSorted([1, 5, 9]));
        $this->assertTrue(isSorted([]));
    }

    public function testSort()
    {
        $empty = [];
        sort($empty);
        $this->assertEquals([], $empty);
        $single = [1];
        sort($single);
        $this->assertEquals([1], $single);

        sort($this->ary);
        $this->assertEquals([1, 2, 4, 7, 8], $this->ary);

        $ary1 = [16, 24, 22, 12];
        sort($ary1);
        $this->assertEquals([12, 16, 22, 24], $ary1);

        $ary2 = [12, 17, 22];
        sort($ary2);
        $this->assertEquals([12, 17, 22], $ary2);
    }

    public function testPartialSort()
    {
        $empty = [];
        partialSort($empty, 1);
        $this->assertEquals([], $empty);

        partialSort($this->ary, 3);
        $this->assertEquals([1, 2, 4], array_slice($this->ary, 0, 3));
        partialSort($this->ary, 4);
        $this->assertEquals([1, 2, 4, 7], array_slice($this->ary, 0, 4));
    }

    public function testPartialSortCopy()
    {
        $this->assertEquals([], partialSortCopy([], 1));

        $this->assertEquals([1, 2], partialSortCopy($this->ary, 2));
        $this->assertEquals([1, 2, 4], partialSortCopy($this->ary, 3));
        $this->assertEquals([1, 2, 4, 7], partialSortCopy($this->ary, 4));
    }

    public function testStableSort()
    {
        $tupleAry = [
            [2, 3],
            [3, 4],
            [3, 2],
            [2, 1],
            [3, 5],
            [2, 4],
            [3, 1],
        ];
        stableSort($tupleAry, function($a, $b) {
            return $a[0] < $b[0];
        });
        $this->assertEquals([
            [2, 3],
            [2, 1],
            [2, 4],
            [3, 4],
            [3, 2],
            [3, 5],
            [3, 1],
        ], $tupleAry);
    }

    public function testNthElement()
    {
        $this->assertEquals(1, nthElement($this->ary, 1));
        $this->assertEquals(2, nthElement($this->ary, 2));
        $this->assertEquals(4, nthElement($this->ary, 3));
        $this->assertEquals(7, nthElement($this->ary, 4));
        $this->assertEquals(8, nthElement($this->ary, 5));
    }
}
?>
